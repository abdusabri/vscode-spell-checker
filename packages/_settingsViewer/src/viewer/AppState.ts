import {observable, computed} from 'mobx';
import { Settings, ConfigTarget, LocalId, SettingByConfigTarget, Config, Configs, LocalList, WorkspaceFolder, TextDocument, configTargets, ConfigTargets } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';


type Maybe<T> = T | undefined;
type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigTarget;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>,
    target: ConfigTarget
}

type InheritedFromTarget<T> = {
    value: Exclude<T, undefined>,
    target: ConfigTarget;
}


type InheritMembers<T> = {
    [K in keyof T]: InheritedFromTarget<T[K]>;
}

type InheritedConfigs = SettingByConfigTarget<Config>;

export class AppState implements State {
    @observable activeTabName = '';
    @observable settings: Settings = {
        dictionaries: [
        ],
        configs: {
            user: { inherited: {}, locals: [], languageIdsEnabled: [] },
            workspace: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            folder: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            file: undefined,
        }
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {}

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter(target => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        return tabs.filter(tab => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find(t => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locals = config.locals; // todo: calc inheritance
            const inherited = config.inherited.locals;

            const infos = new Map<string, LanguageInfo>();

            const addLocalsToInfos = (locals: string[], dictionaryName: string | undefined) => {
                locals.map(normalizeCode).map(lookupCode).filter(notUndefined).forEach(lang => {
                    const { code, lang: language, country } = lang;
                    const name = country ? `${language} - ${country}` : language;
                    const found = this.isLocalEnabledEx(target, code);
                    const enabled = found && found.value || false;
                    const info: LanguageInfo = infos.get(name) || {
                        code,
                        name,
                        enabled,
                        dictionaries: [],
                    };
                    if (dictionaryName) {
                        info.dictionaries.push(dictionaryName);
                    }
                    infos.set(name, info);
                });
            }
            if (locals) {
                addLocalsToInfos(locals, undefined);
            }
            this.settings.dictionaries.forEach(dict => addLocalsToInfos(dict.locals, dict.name));

            return {
                languages: [...infos.values()].sort(compareEach(
                    compareBy(info => !info.dictionaries.length),
                    compareBy('name'),
                )),
                inherited
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        }
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex(t => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = workspace && workspace.workspaceFolders || [];
        return folders.map(f => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.workspaceFolders || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.textDocuments || [];
    }

    @computed get documentSelection(): { label: string, value: string }[] {
        return this.documents.map(doc => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.name === name)[0];
    }

    actionSetLocal(target: ConfigTarget, local: LocalId, enable: boolean) {
        const localsCurrent = this.settings.configs[target].locals;
        const locals = (enable ? [...localsCurrent, local] : localsCurrent.filter(v => v !== local)).filter(uniqueFilter());
        this.settings.configs[target].locals = locals;
        this.messageBus.postMessage({ command: 'EnableLocalMessage', value: { target, local, enable } });
    }

    actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocalId):InheritedFromTarget<boolean> {
        const config = this.settings.configs[field];
        const target = config.inherited.locals || field;
        const locals = config.locals;
        return  {
            value: locals.map(normalizeCode).includes(code),
            target
        };
    }


    actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    actionActivateTab(tabName: string) {
        this.activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri })
        }
    }

    actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (fileConfig && fileConfig.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target }});
    }
}

function notUndefined<T>(a : T): a is Exclude<T, undefined> {
    return a !== undefined;
}


