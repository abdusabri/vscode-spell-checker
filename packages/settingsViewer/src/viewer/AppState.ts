import {observable, computed} from 'mobx';
import { Settings } from '../api/settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable settings: Settings = {
        locals: [
            {
                code: 'en',
                name: 'English',
                dictionaries: ['en', 'en-us'],
                enabled: true,
                isInUserSettings: true,
            },
            {
                code: 'es',
                name: 'Spanish',
                dictionaries: ['es', 'es-ES'],
                enabled: true,
                isInUserSettings: true,
            },
        ],
    };

    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
