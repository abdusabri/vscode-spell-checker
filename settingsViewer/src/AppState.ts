import {observable, computed} from 'mobx';
import { ConfigurationForDocument } from './settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable config: ConfigurationForDocument = {
        languageEnabled: true,
        fileEnabled: true,
        settings: {},
        docSettings: {}
    };
    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
