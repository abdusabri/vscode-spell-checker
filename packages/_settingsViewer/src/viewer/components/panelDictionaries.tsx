import * as React from 'react';
import {Cell, Grid, Row} from '@material/react-layout-grid';
import List, {ListItem, ListItemText, ListItemGraphic, ListItemMeta} from '@material/react-list';
import MaterialIcon from '@material/react-material-icon';
import { AppState } from '../AppState';


export class PanelDictionaries extends React.Component<{appState: AppState}, {}> {
    render() {
        const dictionaries = this.props.appState.settings.dictionaries;
        return (
            <div>
                <h3>Dictionaries</h3>
                <List twoLine className='dictionary_list'>
                    {dictionaries.map(dict => {
                        const hasLocals = dict.locals && dict.locals.length > 0;
                        const hasFileTypes = dict.fileTypes && dict.fileTypes.length > 0;
                        const icon = hasFileTypes
                            ? 'code'
                            : hasLocals ? 'import_contacts'
                            : 'select_all';
                        return (
                        <ListItem key={dict.name}>
                            <ListItemGraphic graphic={<MaterialIcon icon={icon}/>} />
                            <ListItemText primaryText={dict.name} secondaryText={dict.description} />
                            <ListItemMeta meta={dict.locals.join(', ')} />
                        </ListItem>);
                    })}
                </List>
            </div>
        );
     }
}

