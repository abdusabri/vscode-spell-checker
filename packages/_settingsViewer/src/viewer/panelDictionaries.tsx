import * as React from 'react';
import {Cell, Grid, Row} from '@material/react-layout-grid';
import List, {ListItem, ListItemText, ListItemGraphic, ListItemMeta} from '@material/react-list';
import MaterialIcon from '@material/react-material-icon';
import { AppState } from './AppState';


export class PanelDictionaries extends React.Component<{appState: AppState}, {}> {
    render() {
        const dictionaries = this.props.appState.settings.dictionaries;
        // const dict0 = dictionaries.slice(0, Math.floor(dictionaries.length / 3));
        // const dict1 = dictionaries.slice(Math.floor(dictionaries.length / 3), Math.floor(dictionaries.length * 2 / 3));
        // const dict2 = dictionaries.slice(Math.floor(dictionaries.length * 2 / 3));
        const cols = [dictionaries]; // [dict0, dict1, dict2];
        return (
            <Grid>
                <Row>
                    <Cell><h3>Dictionaries</h3></Cell>
                </Row>
                <Row>
                    {cols.map((col, index) =>
                    <Cell columns={8} key={index}>
                        <List twoLine>
                            {col.map(dict =>
                            <ListItem key={dict.name}>
                                <ListItemGraphic graphic={<MaterialIcon icon='import_contacts'/>} />
                                <ListItemText primaryText={dict.name} secondaryText={dict.description} />
                                {/*<ListItemMeta meta={<MaterialIcon icon='check_circle'/>} />*/}
                            </ListItem>
                            )}
                        </List>
                    </Cell>)}
                </Row>
            </Grid>
        );
     }
}

