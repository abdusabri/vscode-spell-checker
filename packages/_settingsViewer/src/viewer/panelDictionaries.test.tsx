
import * as React from 'react';
import { AppState } from './AppState';
import { create } from 'react-test-renderer';
import { sampleSettings } from './samples/sampleSettings';
import { PanelDictionaries } from './panelDictionaries';

describe('Dictionary Panel Verification', () => {
    it('tests the snapshot', () => {
        const appState = getSampleAppState();
        const panelRenderer = create(<PanelDictionaries appState={appState}></PanelDictionaries>).toJSON()!;
        expect(panelRenderer).toMatchSnapshot();
    });

    function getSampleAppState(): AppState {
        const appState: AppState = new AppState();
        appState.settings = sampleSettings;
        return appState;
    }
});
