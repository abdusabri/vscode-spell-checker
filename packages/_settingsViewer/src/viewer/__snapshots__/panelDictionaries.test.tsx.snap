// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary Panel Verification tests the snapshot 1`] = `
<div
  className="mdc-layout-grid"
>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell"
    >
      <h3>
        Dictionaries
      </h3>
    </div>
  </div>
  <div
    className="mdc-layout-grid__inner"
  >
    <div
      className="mdc-layout-grid__cell mdc-layout-grid__cell--span-8"
    >
      <ul
        aria-orientation="vertical"
        className="mdc-list mdc-list--two-line"
      >
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              companies
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Company names dictionary for cspell.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              css
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              CSS Keywords.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              csharp
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              C# Keywords and common library functions.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              dotnet
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              .Net keywords.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              filetypes
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              List of file types.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              fonts
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              List of fonts.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              html
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              HTML keywords.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              misc
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              List of miscellaneous terms.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              node
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              List of NodeJS terms.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              npm
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              List of Top 500 NPM packages.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              powershell
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Powershell Keywords.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              softwareTerms
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Common Software Terms.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              typescript
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              JavaScript and Typescript terms.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              cpp
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              C/C++ Keywords and common library functions.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              django
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              List of Python Django Framework keywords.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              elixir
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Elixir dictionary for cspell.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              en_us
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              American English Dictionary
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              en-gb
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              British English Dictionary
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              fullstack
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Common words encountered during fullstack development
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              golang
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Go Language Dictionary
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              java
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Java dictionary for cspell.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              latex
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              LaTeX dictionary
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              lorem-ipsum
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Lorem-ipsum dictionary for cspell.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              php
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Php dictionary for cspell.
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              python
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Python Keyword Dictionary
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              rust
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Rust Keyword Dictionary
            </span>
          </span>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              scala
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              Scala dictionary for cspell.
            </span>
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
